README.md
# Readme #

[![forthebadge](http://forthebadge.com/images/badges/powered-by-electricity.svg)](http://forthebadge.com)

These files contains my projects for the course CS F422 Parallel Computing by [Dr. Shanmugasundaram Balasubramaniam](http://csis.bits-pilani.ac.in/faculty/sundarb/).
Offered at BITS Pilani during 2nd Semester of 2015-2016.
I am a fourth year undergrad student doing this course.

ass2/news
WORLD 3 Turkey Turkey Detains 10 Suspected ISIS Members In Raids In Bingol.
POLITICS 7 India Bihar Lawmaker Accused Of Raping Minor, Surrenders.
SPORTS 9 England Saina Nehwal Enters Quarterfinals of All England Championship.
SPORTS 9 India Sania Mirza Invites Doubles Partner Martina Hingis to Visit Telangana.
POLITICS 12 India Copernicus, Pythagorus Not Removed From Text Books: Smriti Irani.
POLITICS 12 India Don't Harass Honest Tax Payers, Says Prime Minister Narendra Modi.
WORLD 12 Japan Japan Still Searching For The Missing, 5 Years After Tsunami.
WORLD 12 US Hillary Clinton, Bernie Sanders Find Common Enemy At Presidential Debate.
WORLD 13 Pakistan Pak Sends Memo To India On Fishermen's Early Release.
POLITICS 14 India Goa Government Discredited, AAP May Contest Polls.
SPORTS 22 India ICC World Twenty20: West Bengal Promises Full Security to Pakistan Team.
POLITICS 23 India Not Enough Aircraft To Cover Borders.
POLITICS 23 India Real Estate Bill Passed In Rajya Sabha.
SPORTS 23 India Akashdeep, Sreejesh Get Nominations For Hockey India's Best Player Award.
SPORTS 23 India Mumbai Indians to Play Rising Pune Supergiants on April 9.
SPORTS 23 India Rahil Gangjee Takes Eighth Spot in Asian List.
SPORTS 23 Unknown Lewis Hamilton Reveals Friction With Teammate Nico Rosberg is Real.
SPORTS 23 US Novak Djokovic Spotted With Deepika Padukone in LA Bar.
WORLD 23 Russia Rights Group Attacked In Russia's Caucasus After Assault On Journalists.
SCIENCE 25 India President, PM Congratulate ISRO Scientists On Successful Satellite Launch.
WORLD 27 North-Korea North Korea To Liquidate South Korean Assets.
SPORTS 34 Brazil Brazil Expect Rio Olympics Without Protests Despite Tensions.
SPORTS 34 Unknown Zlatan Ibrahimovic Silenced His Doubters, Says Laurent Blanc.
WORLD 34 9 Lebanon Dead As Lebanese Army Clashes With Terrorists.
WORLD 34 India Vice President Joe Biden Meets With Jordan's King Abdullah.
WORLD 37 Japan Shinzo Abe Says Japan 'Cannot Do Without' Nuclear Power.
POLITICS 39 India On Dance Bar Ban, A Government-Opposition Tango In Maharashtra.
SPORTS 39 India Never Refused Security To India-Pakistan Match, Says Virbhadra Singh.
POLITICS 43 India Amid Sedition Row, JNU Gets President's Award.
WORLD 46 Japan Fukushima's Ground Zero: No Place For Man Or Robot.
WORLD 48 India World's Longest-Serving Warship INS Viraat Looks For A Home.
POLITICS 50 India PM Narendra Modi Congratulates ISRO For Sixth Satellite Launch.
WORLD 54 Saudi-Arabia 'Thunder' In Saudi Desert As Huge Military Drill Ends.
SPORTS 56 US Serena Lauds Sharapova for Owning up to Failed Drug Test.
WORLD 56 South-Africa South African Woman Found Guilty Of 1997 Baby Kidnapping.
POLITICS 65 India Mamata Banerjee Slams Congress-Left 'Rainbow Coalition' In West Bengal.
WORLD 68 US Republican Presidential Candidates May Generate Less Heat At Miami Debate.
SCIENCE 73 US Early Habitat Shows Life Was Stressful For Early Humans.
POLITICS 76 India Sri Sri Event: Swaraj Abhiyan Slams Centre, Delhi Government.
SPORTS 76 India Joost Luiten, Marcus Fraser to Play at Hero Indian Open.
POLITICS 77 India Opposition Corners Government Over Farmers' Suicides.
SPORTS 77 Unknown Anirban Lahiri Placed Tied 11th at World Golf Championship.
WORLD 77 US Indian American Boy In Fray For $100,000 'Child Genius' Prize.
SPORTS 78 India Vijay Mallya Says he is Still The Boss at Force India.
SPORTS 78 Pakistan Pakistan Refuses To Clear Cricket Team's Departure To India For Now.
POLITICS 79 India AGP Comes Out With First List Of Candidates For Assam Polls.
POLITICS 88 India CPI(M) leader Pinarayi Vijayan To Contest Kerala Elections.
SPORTS 88 Unknown FIFA Agent Confession Threatens to Drag Down Others.
SCIENCE 112 Unknown H. erectus cut, chewed way through evolution.
SCIENCE 112 Unknown Ultrasmall engines bend second law of thermodynamics.
SPORTS 112 Unknown Bernie Ecclestone Welcomes Formula 1 Changes After 'Wake-Up' Call.
SCIENCE 122 Unknown Cells from fat mend bone, cartilage, muscle and even the heart.
BUSINESS 123 China China stocks down as inflation data limits room for easing.
POLITICS 123 India Bengal Polls: Election Commission Goes Tech-Savvy.
SCIENCE 123 Unknown Modern Corn Hybrids More Resilient to Nitrogen Stress.
BUSINESS 128 India As Mallya slips out, investors punish banks that have exposure to his firm. 
BUSINESS 128 India Centre to sign pacts worth over Rs.72,000 cr at shipping summit.
BUSINESS 134 India Corporation Bank shares gain over 2.5% on fund raising plans.
BUSINESS 134 India Five-fold Hike In Government Budget For Funding Start-ups.
SCIENCE 134 Unknown Gravity waves detected at last!
SCIENCE 134 Unknown Molecules found to counter antibiotic resistance.
SCIENCE 134 Unknown Neandertal toe contains human DNA.
BUSINESS 136 India Crompton Greaves to sell T&D biz outside India to First Reserve.
BUSINESS 143 India No Signs of Sharp Economic Turnaround in FY17: Crisil.
BUSINESS 145 India Inflation Seen Cooling in February, Gives RBI Easing Room.
SCIENCE 145 Russia Robots welcome visitors to Berlin travel fair.
SCIENCE 145 Unknown Lizards locked in amber provide clues to reptile evolution.
SCIENCE 145 US How a hacker's typo helped stop a billion dollar bank heist.
SCIENCE 156 Unknown Hunt is on for new Planet Nine.
SCIENCE 156 Unknown Teens eating better but gaining weight.
SCIENCE 167 US NASA plans to fix Mars spacecraft leak then launch in 2018.
BUSINESS 182 India Container Corp. stake sale oversubscribed on day one.
BUSINESS 185 India Unichem Laboratories surged 4%.
SPORTS 223 Unknown Michael Schumacher showing 'small signs' of improvement.
SCIENCE 232 Unknown New techniques regrow lens, cornea tissue.
SPORTS 275 Unknown Max Verstappen Set to Become Youngest F1 Driver.


ass1/sha1.c
/*
 *  FIPS-180-1 compliant SHA-1 implementation
 *
 *  Copyright (C) 2006-2015, ARM Limited, All Rights Reserved
 *  SPDX-License-Identifier: Apache-2.0
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"); you may
 *  not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 *  This file is part of mbed TLS (https://tls.mbed.org)
 */
/*
 *  The SHA-1 standard was published by NIST in 1993.
 *
 *  http://www.itl.nist.gov/fipspubs/fip180-1.htm
 */

#if !defined(MBEDTLS_CONFIG_FILE)
#include "include/config.h"
#else
#include MBEDTLS_CONFIG_FILE
#endif

#if defined(MBEDTLS_SHA1_C)

#include "include/sha1.h"

#include <string.h>

#if defined(MBEDTLS_SELF_TEST)
#if defined(MBEDTLS_PLATFORM_C)
#include "include/platform.h"
#else
#include <stdio.h>
#define mbedtls_printf printf
#endif /* MBEDTLS_PLATFORM_C */
#endif /* MBEDTLS_SELF_TEST */

#if !defined(MBEDTLS_SHA1_ALT)

/* Implementation that should never be optimized out by the compiler */
static void mbedtls_zeroize( void *v, size_t n ) {
    volatile unsigned char *p = v; while( n-- ) *p++ = 0;
}

/*
 * 32-bit integer manipulation macros (big endian)
 */
#ifndef GET_UINT32_BE
#define GET_UINT32_BE(n,b,i)                            \
{                                                       \
    (n) = ( (uint32_t) (b)[(i)    ] << 24 )             \
        | ( (uint32_t) (b)[(i) + 1] << 16 )             \
        | ( (uint32_t) (b)[(i) + 2] <<  8 )             \
        | ( (uint32_t) (b)[(i) + 3]       );            \
}
#endif

#ifndef PUT_UINT32_BE
#define PUT_UINT32_BE(n,b,i)                            \
{                                                       \
    (b)[(i)    ] = (unsigned char) ( (n) >> 24 );       \
    (b)[(i) + 1] = (unsigned char) ( (n) >> 16 );       \
    (b)[(i) + 2] = (unsigned char) ( (n) >>  8 );       \
    (b)[(i) + 3] = (unsigned char) ( (n)       );       \
}
#endif

void mbedtls_sha1_init( mbedtls_sha1_context *ctx )
{
    memset( ctx, 0, sizeof( mbedtls_sha1_context ) );
}

void mbedtls_sha1_free( mbedtls_sha1_context *ctx )
{
    if( ctx == NULL )
        return;

    mbedtls_zeroize( ctx, sizeof( mbedtls_sha1_context ) );
}

void mbedtls_sha1_clone( mbedtls_sha1_context *dst,
                         const mbedtls_sha1_context *src )
{
    *dst = *src;
}

/*
 * SHA-1 context setup
 */
void mbedtls_sha1_starts( mbedtls_sha1_context *ctx )
{
    ctx->total[0] = 0;
    ctx->total[1] = 0;

    ctx->state[0] = 0x67452301;
    ctx->state[1] = 0xEFCDAB89;
    ctx->state[2] = 0x98BADCFE;
    ctx->state[3] = 0x10325476;
    ctx->state[4] = 0xC3D2E1F0;
}

#if !defined(MBEDTLS_SHA1_PROCESS_ALT)
void mbedtls_sha1_process( mbedtls_sha1_context *ctx, const unsigned char data[64] )
{
    uint32_t temp, W[16], A, B, C, D, E;

    GET_UINT32_BE( W[ 0], data,  0 );
    GET_UINT32_BE( W[ 1], data,  4 );
    GET_UINT32_BE( W[ 2], data,  8 );
    GET_UINT32_BE( W[ 3], data, 12 );
    GET_UINT32_BE( W[ 4], data, 16 );
    GET_UINT32_BE( W[ 5], data, 20 );
    GET_UINT32_BE( W[ 6], data, 24 );
    GET_UINT32_BE( W[ 7], data, 28 );
    GET_UINT32_BE( W[ 8], data, 32 );
    GET_UINT32_BE( W[ 9], data, 36 );
    GET_UINT32_BE( W[10], data, 40 );
    GET_UINT32_BE( W[11], data, 44 );
    GET_UINT32_BE( W[12], data, 48 );
    GET_UINT32_BE( W[13], data, 52 );
    GET_UINT32_BE( W[14], data, 56 );
    GET_UINT32_BE( W[15], data, 60 );

#define S(x,n) ((x << n) | ((x & 0xFFFFFFFF) >> (32 - n)))

#define R(t)                                            \
(                                                       \
    temp = W[( t -  3 ) & 0x0F] ^ W[( t - 8 ) & 0x0F] ^ \
           W[( t - 14 ) & 0x0F] ^ W[  t       & 0x0F],  \
    ( W[t & 0x0F] = S(temp,1) )                         \
)

#define P(a,b,c,d,e,x)                                  \
{                                                       \
    e += S(a,5) + F(b,c,d) + K + x; b = S(b,30);        \
}

    A = ctx->state[0];
    B = ctx->state[1];
    C = ctx->state[2];
    D = ctx->state[3];
    E = ctx->state[4];

#define F(x,y,z) (z ^ (x & (y ^ z)))
#define K 0x5A827999

    P( A, B, C, D, E, W[0]  );
    P( E, A, B, C, D, W[1]  );
    P( D, E, A, B, C, W[2]  );
    P( C, D, E, A, B, W[3]  );
    P( B, C, D, E, A, W[4]  );
    P( A, B, C, D, E, W[5]  );
    P( E, A, B, C, D, W[6]  );
    P( D, E, A, B, C, W[7]  );
    P( C, D, E, A, B, W[8]  );
    P( B, C, D, E, A, W[9]  );
    P( A, B, C, D, E, W[10] );
    P( E, A, B, C, D, W[11] );
    P( D, E, A, B, C, W[12] );
    P( C, D, E, A, B, W[13] );
    P( B, C, D, E, A, W[14] );
    P( A, B, C, D, E, W[15] );
    P( E, A, B, C, D, R(16) );
    P( D, E, A, B, C, R(17) );
    P( C, D, E, A, B, R(18) );
    P( B, C, D, E, A, R(19) );

#undef K
#undef F

#define F(x,y,z) (x ^ y ^ z)
#define K 0x6ED9EBA1

    P( A, B, C, D, E, R(20) );
    P( E, A, B, C, D, R(21) );
    P( D, E, A, B, C, R(22) );
    P( C, D, E, A, B, R(23) );
    P( B, C, D, E, A, R(24) );
    P( A, B, C, D, E, R(25) );
    P( E, A, B, C, D, R(26) );
    P( D, E, A, B, C, R(27) );
    P( C, D, E, A, B, R(28) );
    P( B, C, D, E, A, R(29) );
    P( A, B, C, D, E, R(30) );
    P( E, A, B, C, D, R(31) );
    P( D, E, A, B, C, R(32) );
    P( C, D, E, A, B, R(33) );
    P( B, C, D, E, A, R(34) );
    P( A, B, C, D, E, R(35) );
    P( E, A, B, C, D, R(36) );
    P( D, E, A, B, C, R(37) );
    P( C, D, E, A, B, R(38) );
    P( B, C, D, E, A, R(39) );

#undef K
#undef F

#define F(x,y,z) ((x & y) | (z & (x | y)))
#define K 0x8F1BBCDC

    P( A, B, C, D, E, R(40) );
    P( E, A, B, C, D, R(41) );
    P( D, E, A, B, C, R(42) );
    P( C, D, E, A, B, R(43) );
    P( B, C, D, E, A, R(44) );
    P( A, B, C, D, E, R(45) );
    P( E, A, B, C, D, R(46) );
    P( D, E, A, B, C, R(47) );
    P( C, D, E, A, B, R(48) );
    P( B, C, D, E, A, R(49) );
    P( A, B, C, D, E, R(50) );
    P( E, A, B, C, D, R(51) );
    P( D, E, A, B, C, R(52) );
    P( C, D, E, A, B, R(53) );
    P( B, C, D, E, A, R(54) );
    P( A, B, C, D, E, R(55) );
    P( E, A, B, C, D, R(56) );
    P( D, E, A, B, C, R(57) );
    P( C, D, E, A, B, R(58) );
    P( B, C, D, E, A, R(59) );

#undef K
#undef F

#define F(x,y,z) (x ^ y ^ z)
#define K 0xCA62C1D6

    P( A, B, C, D, E, R(60) );
    P( E, A, B, C, D, R(61) );
    P( D, E, A, B, C, R(62) );
    P( C, D, E, A, B, R(63) );
    P( B, C, D, E, A, R(64) );
    P( A, B, C, D, E, R(65) );
    P( E, A, B, C, D, R(66) );
    P( D, E, A, B, C, R(67) );
    P( C, D, E, A, B, R(68) );
    P( B, C, D, E, A, R(69) );
    P( A, B, C, D, E, R(70) );
    P( E, A, B, C, D, R(71) );
    P( D, E, A, B, C, R(72) );
    P( C, D, E, A, B, R(73) );
    P( B, C, D, E, A, R(74) );
    P( A, B, C, D, E, R(75) );
    P( E, A, B, C, D, R(76) );
    P( D, E, A, B, C, R(77) );
    P( C, D, E, A, B, R(78) );
    P( B, C, D, E, A, R(79) );

#undef K
#undef F

    ctx->state[0] += A;
    ctx->state[1] += B;
    ctx->state[2] += C;
    ctx->state[3] += D;
    ctx->state[4] += E;
}
#endif /* !MBEDTLS_SHA1_PROCESS_ALT */

/*
 * SHA-1 process buffer
 */
void mbedtls_sha1_update( mbedtls_sha1_context *ctx, const unsigned char *input, size_t ilen )
{
    size_t fill;
    uint32_t left;

    if( ilen == 0 )
        return;

    left = ctx->total[0] & 0x3F;
    fill = 64 - left;

    ctx->total[0] += (uint32_t) ilen;
    ctx->total[0] &= 0xFFFFFFFF;

    if( ctx->total[0] < (uint32_t) ilen )
        ctx->total[1]++;

    if( left && ilen >= fill )
    {
        memcpy( (void *) (ctx->buffer + left), input, fill );
        mbedtls_sha1_process( ctx, ctx->buffer );
        input += fill;
        ilen  -= fill;
        left = 0;
    }

    while( ilen >= 64 )
    {
        mbedtls_sha1_process( ctx, input );
        input += 64;
        ilen  -= 64;
    }

    if( ilen > 0 )
        memcpy( (void *) (ctx->buffer + left), input, ilen );
}

static const unsigned char sha1_padding[64] =
{
 0x80, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/*
 * SHA-1 final digest
 */
void mbedtls_sha1_finish( mbedtls_sha1_context *ctx, unsigned char output[20] )
{
    uint32_t last, padn;
    uint32_t high, low;
    unsigned char msglen[8];

    high = ( ctx->total[0] >> 29 )
         | ( ctx->total[1] <<  3 );
    low  = ( ctx->total[0] <<  3 );

    PUT_UINT32_BE( high, msglen, 0 );
    PUT_UINT32_BE( low,  msglen, 4 );

    last = ctx->total[0] & 0x3F;
    padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );

    mbedtls_sha1_update( ctx, sha1_padding, padn );
    mbedtls_sha1_update( ctx, msglen, 8 );

    PUT_UINT32_BE( ctx->state[0], output,  0 );
    PUT_UINT32_BE( ctx->state[1], output,  4 );
    PUT_UINT32_BE( ctx->state[2], output,  8 );
    PUT_UINT32_BE( ctx->state[3], output, 12 );
    PUT_UINT32_BE( ctx->state[4], output, 16 );
}

#endif /* !MBEDTLS_SHA1_ALT */

/*
 * output = SHA-1( input buffer )
 */
void mbedtls_sha1( const unsigned char *input, size_t ilen, unsigned char output[20] )
{
    mbedtls_sha1_context ctx;

    mbedtls_sha1_init( &ctx );
    mbedtls_sha1_starts( &ctx );
    mbedtls_sha1_update( &ctx, input, ilen );
    mbedtls_sha1_finish( &ctx, output );
    mbedtls_sha1_free( &ctx );
}

#if defined(MBEDTLS_SELF_TEST)
/*
 * FIPS-180-1 test vectors
 */
static const unsigned char sha1_test_buf[3][57] =
{
    { "abc" },
    { "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq" },
    { "" }
};

static const int sha1_test_buflen[3] =
{
    3, 56, 1000
};

static const unsigned char sha1_test_sum[3][20] =
{
    { 0xA9, 0x99, 0x3E, 0x36, 0x47, 0x06, 0x81, 0x6A, 0xBA, 0x3E,
      0x25, 0x71, 0x78, 0x50, 0xC2, 0x6C, 0x9C, 0xD0, 0xD8, 0x9D },
    { 0x84, 0x98, 0x3E, 0x44, 0x1C, 0x3B, 0xD2, 0x6E, 0xBA, 0xAE,
      0x4A, 0xA1, 0xF9, 0x51, 0x29, 0xE5, 0xE5, 0x46, 0x70, 0xF1 },
    { 0x34, 0xAA, 0x97, 0x3C, 0xD4, 0xC4, 0xDA, 0xA4, 0xF6, 0x1E,
      0xEB, 0x2B, 0xDB, 0xAD, 0x27, 0x31, 0x65, 0x34, 0x01, 0x6F }
};

/*
 * Checkup routine
 */
int mbedtls_sha1_self_test( int verbose )
{
    int i, j, buflen, ret = 0;
    unsigned char buf[1024];
    unsigned char sha1sum[20];
    mbedtls_sha1_context ctx;

    mbedtls_sha1_init( &ctx );

    /*
     * SHA-1
     */
    for( i = 0; i < 3; i++ )
    {
        if( verbose != 0 )
            mbedtls_printf( "  SHA-1 test #%d: ", i + 1 );

        mbedtls_sha1_starts( &ctx );

        if( i == 2 )
        {
            memset( buf, 'a', buflen = 1000 );

            for( j = 0; j < 1000; j++ )
                mbedtls_sha1_update( &ctx, buf, buflen );
        }
        else
            mbedtls_sha1_update( &ctx, sha1_test_buf[i],
                               sha1_test_buflen[i] );

        mbedtls_sha1_finish( &ctx, sha1sum );

        if( memcmp( sha1sum, sha1_test_sum[i], 20 ) != 0 )
        {
            if( verbose != 0 )
                mbedtls_printf( "failed\n" );

            ret = 1;
            goto exit;
        }

        if( verbose != 0 )
            mbedtls_printf( "passed\n" );
    }

    if( verbose != 0 )
        mbedtls_printf( "\n" );

exit:
    mbedtls_sha1_free( &ctx );

    return( ret );
}

#endif /* MBEDTLS_SELF_TEST */

#endif /* MBEDTLS_SHA1_C */
